libcache/cache.c

```diff
@@ -0,0 +1,391 @@
+				break;
+			}
+		}
+
+		temp = &(set->lines[i]);
+
+		switch (policy) {
+			case 0: {
+				/* write-back */
+				set->lines[i].isDirty = '1';
+				break;
+			}
+			case 1: {
+				/* write-through */
+				wrt = (*writeCb)(offset, temp->data, sizeof *(temp->data), 1);
+				break;
+			}
+			default: {
+				break;
+			}
+		}
+
+		for (i = 0; i < LIBCACHE_NUM_WAYS; ++i) {
+			if (set->tags[i] == NULL) {
+				set->tags[i] = temp;
+				break;
+			}
+		}
+
+		LIST_ADD(&set->timestamps, temp);
+		set->count++;
+	}
+	else {
+		oldest = set->timestamps;
+		LIST_REMOVE(&set->timestamps, oldest);
+
+		/* write-back */
+		if (oldest->isDirty == '1') {
+			wrt = (*writeCb)(offset, oldest->data, sizeof *(oldest->data), 1);
+		}
+
+		for (; i < LIBCACHE_NUM_WAYS; ++i) {
+			if (set->lines[i].tag == oldest->tag) {
+				set->lines[i] = *line;
+				break;
+			}
+		}
+
+		switch (policy) {
+			case 0: {
+				/* write-back */
+				set->lines[i].isDirty = '1';
+				break;
+			}
+			case 1: {
+				/* write-through */
+				wrt = (*writeCb)(offset, oldest->data, sizeof *(oldest->data), 1);
+				break;
+			}
+			default: {
+				break;
+			}
+		}
+
+		LIST_ADD(&set->timestamps, oldest);
+	}
+	qsort(set->tags, LIBCACHE_NUM_WAYS, sizeof(cacheline_t *), cache_compareTags);
+
+	return wrt;
+}
+
+
+uint32_t *cache_readFromSet(cacheset_t *set, uint64_t tag, off_t offset)
+{
+	uint32_t *ret = NULL;
+	cacheline_t key, *keyPtr = &key, **linePtr = NULL;
+
+	key.tag = tag;
+
+	linePtr = bsearch(&keyPtr, set->tags, LIBCACHE_NUM_WAYS, sizeof(cacheline_t *), cache_compareTags);
+
+	if (linePtr != NULL) {
+		ret = (*linePtr)->data;
+
+		LIST_REMOVE(&set->timestamps, (*linePtr));
+		LIST_ADD(&set->timestamps, *linePtr);
+	}
+
+	return ret;
+}
+
+
+ssize_t cache_write(cachectx_t *cache, const off_t addr, void *buffer, int policy)
+{
+	ssize_t wrt = 0;
+	uint32_t *data = (uint32_t *)buffer;
+
+	off_t setIndex = (addr >> cache->offBitNum) & cache->setMask;
+	off_t offset = addr & cache->offMask;
+	off_t tag = (addr & cache->tagMask);
+
+	printf("set index: %lu\n", setIndex);
+	printf("addr: %lu\n", addr);
+	printf("tag: %lu\n", tag);
+
+	cacheline_t line = { tag, data, '1', '0' };
+
+	return cache_writeToSet(cache->writeCb, &cache->sets[setIndex], &line, offset, policy);
+}
+
+
+ssize_t cache_read(cachectx_t *cache, const off_t addr, void *buffer)
+{
+	uint32_t *data = (uint32_t *)buffer, *ret;
+	ssize_t wrt = 0;
+
+	off_t setIndex = (addr >> cache->offBitNum) & cache->setMask;
+	off_t offset = addr & cache->offMask;
+	off_t tag = (addr & cache->tagMask);
+
+	printf("set index: %lu\n", setIndex);
+	printf("offset: %lu\n", offset);
+	printf("tag: %lu\n", tag);
+
+	ret = cache_readFromSet(&cache->sets[setIndex], tag, offset);
+
+	/* cache miss */
+	if (ret == NULL) {
+		cache->readCb(offset, data, sizeof *data, 1);
+		printf("data read by callback: %u\n", *data);
+
+		wrt = cache_write(cache, addr, data, -1);
+
+		ret = cache_readFromSet(&cache->sets[setIndex], tag, offset);
```

Comment:
No need - we already have data we need in `data`

Reply:
True, thank you.